pss/pss.go

```diff
@@ -59,13 +68,37 @@ func (p *Pss) Send(ctx context.Context, targets [][]byte, topic trojan.Topic, pa
 		return nil, err
 	}
 
+	tag, err := p.tags.Create("pss-chunks-tag", 1, false)
+	if err != nil {
+		return nil, err
+	}
+
 	// SAVE trojanChunk to localstore, if it exists do nothing as it's already peristed
-	// TODO: for second phase, use tags --> listen for response of recipient, recipient offline
-	if _, err = p.localStore.Put(ctx, chunk.ModePutUpload, tc); err != nil {
+	if _, err = p.localStore.Put(ctx, chunk.ModePutUpload, tc.WithTagID(tag.Uid)); err != nil {
 		return nil, err
 	}
+	tag.Total = 1
```

Comment:
When the Send method is called a new tag is created each time `tag, err := p.tags.Create("pss-chunks-tag", 1, false)`, the name is the same but the uID is not, this is what really identifies a tag.

Because we are using `tags.Create` this is using the under the hood a random ID for each tag.

```
// Create creates a new tag, stores it by the name and returns it
// it returns an error if the tag with this name already exists
func (ts *Tags) Create(s string, total int64, anon bool) (*Tag, error) {
	t := NewTag(TagUidFunc(), s, total, anon)

	if _, loaded := ts.tags.LoadOrStore(t.Uid, t); loaded {
		return nil, errExists
	}

	return t, nil
}
```
This uses `TagUidFunc` as the uID for the tag
```
var (
	TagUidFunc     = rand.Uint32
	TagNotFoundErr = errors.New("tag not found")
)

```

> if a node calls Send twice, then the pss-chunks-tag tag will already exist, no?

This will not result in a duplicate tag, just multiple tags in the store with the same name but with different IDs, so basically no collision will happen, another story would be if we used a constant id for all the tags then this would happen for sure, this is not the case.

> should we not just proceed with the already existing tag?
As it's creating a new tag it will continue without a problem and the `Send` method will work just fine.

Here is a test you can do, to verify 2 unique tags are generated each time
Modify `TestTrojanChunkRetrieval` and add this after the `Send` method
```
       if _, err = pss.Send(ctx, targets, topic, payload); err != nil {
		t.Fatal(err)
	}

	storeTags := tags.All()
	if len(storeTags) != 2 {
		t.Fatalf("expected %d tags got %d", 2, len(storeTags))
	}
```

Reply:
> also: could we have 2 concurrent Send calls which both set tag.Total to 1? would we have a problem in that case?

The total field is only applied to each individual tag, so each tag will indeed have one, this is not a problem as there will be multiple tags each with a Total of 1

Reply:
that clears it up. thanks for the very thorough explanation!